src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,914 @@
+    (context.chapter === 1 && primitiveType === 'null')
+  ) {
+    context.errors.push(new TypeNotAllowedError(node, primitiveType))
+    return tAny
+  }
+  return tPrimitive(primitiveType as PrimitiveType | TSAllowedTypes)
+}
+
+/**
+ * Wrapper function for lookupType that removes ForAll and Predicate types,
+ * since they are not used in the type error checker.
+ */
+function lookupTypeAndRemoveForAllAndPredicateTypes(
+  name: string,
+  env: TypeEnvironment
+): Type | undefined {
+  const type = lookupType(name, env)
+  if (!type) {
+    return undefined
+  }
+  if (type.kind === 'forall') {
+    // Skip typecheck as function has variable number of arguments;
+    // this only occurs for certain prelude functions
+    return tAny
+  }
+  if (type.kind === 'predicate') {
+    // All predicate functions (e.g. is_number)
+    // take in 1 parameter and return a boolean
+    return tFunc(tAny, tBool)
+  }
+  return type
+}
+
+/**
+ * Combines all types provided in the parameters into one, removing duplicate types in the process.
+ */
+function mergeTypes(...types: Type[]): Type {
+  const mergedTypes: Type[] = []
+  for (const currType of types) {
+    if (isEqual(currType, tAny)) {
+      return tAny
+    }
+    if (currType.kind === 'union') {
+      for (const type of currType.types) {
+        if (!containsType(mergedTypes, type)) {
+          mergedTypes.push(type)
+        }
+      }
+    } else if (!containsType(mergedTypes, currType)) {
+      mergedTypes.push(currType)
+    } else {
+      // Duplicate type, do nothing
+    }
+  }
+  if (mergedTypes.length === 1) {
+    return mergedTypes[0]
+  }
+  return tUnion(...mergedTypes)
+}
+
+/**
+ * Checks if a type exists in an array of types.
+ */
+function containsType(arr: Type[], typeToCheck: Type) {
+  for (const type of arr) {
+    if (isEqual(type, typeToCheck)) {
+      return true
+    }
+    if (typeToCheck.kind === 'primitive') {
+      // If both types are primitives, ignore values
+      if (type.kind === 'primitive' && typeToCheck.name === type.name) {
+        return true
+      }
+      // If checking primitive against literal, check by value
+      if (type.kind === 'literal' && typeToCheck.value === type.value) {
+        return true
+      }
+    }
+    // If checking literal against primitive, check by type
+    if (
+      typeToCheck.kind === 'literal' &&
+      type.kind === 'primitive' &&
+      typeof typeToCheck.value === type.name
+    ) {
+      return true
+    }
+  }
+  return false
+}
+
+/**
+ * Traverses through the program and removes all TS-related nodes, returning the result.
+ */
+function removeTSNodes(node: tsEs.Node): any {
+  const type = node.type
+  switch (type) {
+    case 'Literal':
+    case 'Identifier': {
+      return node
+    }
+    case 'Program':
+    case 'BlockStatement': {
+      const newBody: tsEs.Statement[] = []
+      node.body.forEach(stmt => {
+        const type = stmt.type
+        if (type.startsWith('TS')) {
```

Comment:
Minor nit: This code can be made more robust to future changes by declaring some container containing all the types you want to match against. For example, if I were to add a new type that starts with `TS`, but is not meant to be matched here, I wouldn't know that it will be treated differently just by looking at the types.